File: backend/domain/entities/purchase_order.py
from typing import List
from domain.enums.category import Category
from domain.enums.unit_type import UnitType
from domain.enums.vehicle_type import VehicleType
from domain.entities.vehicle import Vehicle
from .product import Product
from domain.enums.uf import UF 

class PurchaseOrder:
    def __init__(self, order_number: str, customer_name: str, city: str, uf: UF, created_by_id: str, total_volume_m3: float):
        self._order_number = order_number
        self._customer_name = customer_name
        self._city = city
        self._uf = uf
        self._created_by_id = created_by_id
        self._total_volume_m3 = total_volume_m3
        self._total_freight = 0.0
        self._items: List[Product] = []
        self._vehicles: List[Vehicle] = []

    @property
    def order_number(self) -> str:
        return self._order_number
    
    @order_number.setter
    def order_number(self, value: str):
        self._order_number = value

    @property
    def customer_name(self) -> str:
        return self._customer_name
    
    @customer_name.setter
    def customer_name(self, value: str):
        self._customer_name = value

    @property
    def city(self) -> str:
        return self._city
    
    @city.setter
    def city(self, value: str):
        self._city = value

    @property
    def uf(self) -> UF:
        return self._uf
    
    @uf.setter
    def uf(self, value: UF):
        self._uf = value
    
    @property
    def created_by_id(self) -> str:
        return self._created_by_id
    
    @created_by_id.setter
    def created_by_id(self, value: str):
        self._created_by_id = value
    
    @property
    def total_volume_m3(self) -> float:
        return self._total_volume_m3
    
    @total_volume_m3.setter
    def total_volume_m3(self, value: float):
        if value < 0:
            raise ValueError("O volume total não pode ser negativo")
        self._total_volume_m3 = value

    @property
    def total_freight(self) -> float:
        return self._total_freight

    @total_freight.setter
    def total_freight(self, value: float):
        self._total_freight = round(value, 2)

    @property
    def items(self) -> List[Product]:
        return self._items
    
    @property
    def vehicles(self) -> List[Vehicle]:
        return self._vehicles 

    def add_item(self, product: Product):
        self._items.append(product)
    
    def add_vehicle(self, vehicle: Vehicle):
        self._vehicles.append(vehicle)

    @property
    def total_products_quantity(self) -> int:
        return sum(item.quantity for item in self._items)
    
    #capacity_ref 60 -> Carreta
    #capacity_ref 45 -> Container/truck
    def get_linear_meters(self, capacity_ref: float = 60.0) -> float:
        if self.total_volume_m3 == 0:
            return 0.0
        return (self.total_volume_m3 * 12) / capacity_ref

    def get_meters_nvia(self, capacity_ref: float = 60.0) -> float:
        return round(self.get_linear_meters(capacity_ref) * 1.10, 2)

    def get_meters_venda(self, capacity_ref: float = 60.0) -> float:
        return round(self.get_linear_meters(capacity_ref) * 1.20, 2)
    
    def to_dict(self):
        return {
            "order_number": self.order_number,
            "customer_name": self.customer_name,
            "city": self.city,
            "uf": self.uf.abbreviation,
            "created_by_id": self.created_by_id,
            "total_volume_m3": self.total_volume_m3,
            "total_freight": self.total_freight,
            "items": [
                {
                    "code": item.code,
                    "description": item.description,
                    "quantity": item.quantity,
                    "unit": item.unit.name,
                    "category": item.category.value
                } for item in self.items
            ],
            "vehicles": [v.to_dict() for v in self.vehicles]
        }
    
    @staticmethod
    def from_dict(data: dict) -> 'PurchaseOrder':

        order = PurchaseOrder(
            order_number=data["order_number"],
            customer_name=data["customer_name"],
            city=data["city"],
            uf=UF[data["uf"]],
            created_by_id=data["created_by_id"],
            total_volume_m3=float(data["total_volume_m3"]),
            total_freight=float(data.get("total_freight", 0.0))
        )

        for item in data.get("items", []):
            product = Product(
                code=item["code"],
                description=item["description"],
                quantity=int(item["quantity"]),
                unit=UnitType(item["unit"]),
                category=Category(item["category"])
            )
            order.add_item(product)

        for v in data.get("vehicles", []):
            vehicle = Vehicle(
                capacity_m3=float(v["capacity_m3"]),
                type=VehicleType(v["type"]),
                quantity=int(v["quantity"])
            )
            order.add_vehicle(vehicle)

        return order

File: frontend/src/domain/entities/PurchaseOrder.ts
import { Product } from "./Product";
import { Vehicle } from "./Vehicle";
import { getUFAbbreviation, UF } from "../enums/UF";
import { VehicleType } from "../enums/VehicleType";
import { UnitType } from "../enums/UnitType";
import { Category } from "../enums/Category";

interface ProductJson {
  code: string;
  description: string;
  quantity: number;
  unit: string;
  category: string;
}

interface VehicleJson {
  capacity_m3: number;
  type: VehicleType;
  quantity: number;
}

interface PurchaseOrderJson {
  order_number: string;
  customer_name: string;
  city: string;
  uf: string;
  created_by_id: string;
  total_volume_m3: number;
  total_freight: number;
  items: ProductJson[];
  vehicles: VehicleJson[];
}

export type ApiDataPayload = Record<
  string, 
  string | number | boolean | ProductJson[] | VehicleJson[] | null | undefined
>;

export class PurchaseOrder {
  private orderNumberAttr: string;
  private customerNameAttr: string;
  private cityAttr: string;
  private ufAttr: UF;
  private createdByIdAttr: string;
  private totalVolumeM3Attr: number;
  private totalFreightAttr: number;
  private itemsAttr: Product[];
  private vehiclesAttr: Vehicle[];

  constructor();

  constructor(
    order_number: string,
    customer_name: string,
    city: string,
    uf: UF,
    created_by_id: string,
    total_volume_m3: number,
    total_freight?: number
  );

  constructor(
    order_number?: string,
    customer_name?: string,
    city?: string,
    uf?: UF,
    created_by_id?: string,
    total_volume_m3?: number,
    total_freight: number = 0
  ) {
    this.orderNumberAttr = order_number ?? "";
    this.customerNameAttr = customer_name ?? "";
    this.cityAttr = city ?? "";
    this.ufAttr = uf!;
    this.createdByIdAttr = created_by_id ?? "";
    this.totalVolumeM3Attr = total_volume_m3 ?? 0;
    this.totalFreightAttr = total_freight;
    this.itemsAttr = [];
    this.vehiclesAttr = [];

    if (total_volume_m3 !== undefined && total_volume_m3 < 0) {
      throw new Error("O volume total não pode ser negativo");
    }
  }

  public get order_number(): string { return this.orderNumberAttr; }
  public set order_number(value: string) { this.orderNumberAttr = value; }

  public get customer_name(): string { return this.customerNameAttr; }
  public set customer_name(value: string) { this.customerNameAttr = value; }

  public get city(): string { return this.cityAttr; }
  public set city(value: string) { this.cityAttr = value; }

  public get uf(): UF { return this.ufAttr; }
  public set uf(value: UF) { this.ufAttr = value; }

  public get created_by_id(): string { return this.createdByIdAttr; }
  public set created_by_id(value: string) { this.createdByIdAttr = value; }

  public get total_volume_m3(): number { return this.totalVolumeM3Attr; }
  public set total_volume_m3(value: number) {
    if (value < 0) {
      throw new Error("O volume total não pode ser negativo");
    }
    this.totalVolumeM3Attr = value;
  }

  public get total_freight(): number { return this.totalFreightAttr; }
  public set total_freight(value: number) {
    this.totalFreightAttr = Number(value.toFixed(2));
  }

  public get items(): Product[] { return this.itemsAttr; }
  public get vehicles(): Vehicle[] { return this.vehiclesAttr; }

  public add_item(product: Product): void {
    this.itemsAttr.push(product);
  }

  public add_vehicle(vehicle: Vehicle): void {
    this.vehiclesAttr.push(vehicle);
  }

  public get total_products_quantity(): number {
    return this.itemsAttr.reduce((sum, item) => sum + item.quantity, 0);
  }

  public get_linear_meters(capacity_ref: number = 60.0): number {
    if (this.total_volume_m3 === 0) {
      return 0.0;
    }
    return (this.total_volume_m3 * 12) / capacity_ref;
  }

  public get_meters_nvia(capacity_ref: number = 60.0): number {
    const meters = this.get_linear_meters(capacity_ref) * 1.10;
    return Number(meters.toFixed(2));
  }

  public get_meters_venda(capacity_ref: number = 60.0): number {
    const meters = this.get_linear_meters(capacity_ref) * 1.20;
    return Number(meters.toFixed(2));
  }

  public toDict(): PurchaseOrderJson {
    return {
      order_number: this.order_number,
      customer_name: this.customer_name,
      city: this.city,
      uf: getUFAbbreviation(this.uf),
      created_by_id: this.created_by_id,
      total_volume_m3: this.total_volume_m3,
      total_freight: this.total_freight,
      items: this.items.map((item) => ({
        code: item.code,
        description: item.description,
        quantity: item.quantity,
        unit: String(item.unit),
        category: String(item.category),
      })),
      vehicles: this.vehicles.map((v) => v.toDict() as unknown as Vehicle),
    };
  }

  static fromDict(data: ApiDataPayload): PurchaseOrder {
    const order = new PurchaseOrder(
      String(data.order_number ?? ""),
      String(data.customer_name ?? ""),
      String(data.city ?? ""),
      data.uf as UF,
      String(data.created_by_id ?? ""),
      Number(data.total_volume_m3 ?? 0),
      Number(data.total_freight ?? 0.0)
    );

    if (data.items && Array.isArray(data.items)) {
      (data.items as ProductJson[]).forEach((item) => {
        const product = new Product(
          item.code,
          item.description,
          Number(item.quantity),
          item.unit as UnitType,
          item.category as Category
        );
        order.add_item(product);
      });
    }

    if (data.vehicles && Array.isArray(data.vehicles)) {
      (data.vehicles as VehicleJson[]).forEach((v) => {
        const vehicle = new Vehicle(
          Number(v.capacity_m3),
          v.type,
          Number(v.quantity)
        );
        order.add_vehicle(vehicle);
      });
    }

    return order;
  }
}